feature/platform/transceiver/tests/zr_firmware_version_test/zr_firmware_version_test.go

```diff
@@ -0,0 +1,148 @@
+	"github.com/openconfig/ygot/ygot"
+)
+
+const (
+	dp16QAM           = 1
+	targetOutputPower = -10
+	frequency         = 193100000
+)
+
+func TestMain(m *testing.M) {
+	fptest.RunTests(m)
+}
+
+// Topology: dut:port1 <--> port2:dut
+
+func configInterface(t *testing.T, dut1 *ondatra.DUTDevice, dp *ondatra.Port, enable bool) {
+	d := &oc.Root{}
+	i := d.GetOrCreateInterface(dp.Name())
+	i.Enabled = ygot.Bool(enable)
+	i.Type = oc.IETFInterfaces_InterfaceType_ethernetCsmacd
+	gnmi.Replace(t, dut1, gnmi.OC().Interface(dp.Name()).Config(), i)
+	component := opticalChannelComponentFromPort(t, dut1, dp)
+	gnmi.Replace(t, dut1, gnmi.OC().Component(component).OpticalChannel().Config(), &oc.Component_OpticalChannel{
+		TargetOutputPower: ygot.Float64(targetOutputPower),
+		Frequency:         ygot.Uint64(frequency),
+	})
+}
+func verifyFirmwareVersionValue(t *testing.T, dut1 *ondatra.DUTDevice, pStream *samplestream.SampleStream[string]) {
+	firmwareVersionSample := pStream.Next(t)
+	if firmwareVersionSample == nil {
+		t.Fatalf("Firmware telemetry %v was not streamed in the most recent subscription interval", firmwareVersionSample)
+	}
+	firmwareVersionVal, ok := firmwareVersionSample.Val()
+	if !ok {
+		t.Fatalf("Firmware version %q telemetry is not present", firmwareVersionSample)
+	}
+	// Check firmware version return value of correct type
+	if reflect.TypeOf(firmwareVersionVal).Kind() != reflect.String {
+		t.Fatalf("Return value is not type string")
+	}
+}
+
+func TestZRFirmwareVersionState(t *testing.T) {
+	dut1 := ondatra.DUT(t, "dut")
+	dp1 := dut1.Port(t, "port1")
+	dp2 := dut1.Port(t, "port2")
+	t.Logf("dut1: %v", dut1)
+	t.Logf("dut1 dp1 name: %v", dp1.Name())
+	configInterface(t, dut1, dp1, true)
+	configInterface(t, dut1, dp2, true)
+	gnmi.Await(t, dut1, gnmi.OC().Interface(dp1.Name()).OperStatus().State(), time.Minute, oc.Interface_OperStatus_UP)
+	transceiverName := gnmi.Get(t, dut1, gnmi.OC().Interface(dp1.Name()).Transceiver().State())
+	// Check if TRANSCEIVER is of type 400ZR
+	if dp1.PMD() != ondatra.PMD400GBASEZR {
+		t.Fatalf("%s Transceiver is not 400ZR its of type: %v", transceiverName, dp1.PMD())
+	}
+	component1 := gnmi.OC().Component(transceiverName)
+
+	p1Stream := samplestream.New(t, dut1, component1.FirmwareVersion().State(), 10*time.Second)
+
+	verifyFirmwareVersionValue(t, dut1, p1Stream)
+
+	p1Stream.Close()
+}
+func TestZRFirmwareVersionStateInterfaceFlap(t *testing.T) {
+	dut1 := ondatra.DUT(t, "dut")
+	dp1 := dut1.Port(t, "port1")
+	dp2 := dut1.Port(t, "port2")
+	t.Logf("dut1: %v", dut1)
+	t.Logf("dut1 dp1 name: %v", dp1.Name())
+	configInterface(t, dut1, dp1, true)
+	configInterface(t, dut1, dp2, true)
+	gnmi.Await(t, dut1, gnmi.OC().Interface(dp1.Name()).OperStatus().State(), time.Minute, oc.Interface_OperStatus_UP)
+	transceiverName := gnmi.Get(t, dut1, gnmi.OC().Interface(dp1.Name()).Transceiver().State())
+	// Check if TRANSCEIVER is of type 400ZR
+	if dp1.PMD() != ondatra.PMD400GBASEZR {
```

Comment:
should this be moved to line 97?  If we didn't get the intended hardware, no sense in configuring and getting the name (lines 99+)?